internal/raft/append_entries.go

```diff
@@ -6,6 +6,50 @@ import "github.com/tomarrell/lbadd/internal/raft/message"
 // to the follower node. This function generates the response to be sent to the leader node.
 // This is the response to the contact by the leader to assert it's leadership.
 func AppendEntriesResponse(node *Node, req *message.AppendEntriesRequest) *message.AppendEntriesResponse {
+	success := true
+	leaderTerm := req.GetTerm()
+	nodePersistentState := node.PersistentState
+	nodeTerm := nodePersistentState.CurrentTerm
+	if nodeTerm > leaderTerm {
+		success = false
+	} else if req.GetPrevLogIndex() > int32(node.VolatileState.CommitIndex) {
+		success = false
+	} else if nodePersistentState.Log[req.PrevLogIndex].Term != req.GetPrevLogTerm() {
+		success = false
+	}
+
+	if !success {
+		return &message.AppendEntriesResponse{
+			Term:    nodeTerm,
+			Success: success,
+		}
+	}
+
+	entries := req.GetEntries()
+	if len(entries) > 0 { // if heartbeat, skip adding entries
+		nodePersistentState.mu.Lock()
+		if req.GetPrevLogIndex() < int32(node.VolatileState.CommitIndex) {
+			node.PersistentState.Log = node.PersistentState.Log[:req.GetPrevLogIndex()]
+		}
+		for _, entry := range entries {
+			node.PersistentState.Log = append(node.PersistentState.Log, *entry)
+		}
+		node.PersistentState.mu.Unlock()
+	}
+
+	if req.GetLeaderCommit() > int32(node.VolatileState.CommitIndex) {
+		nodeCommitIndex := int(req.GetLeaderCommit())
+		if int(req.GetLeaderCommit()) > len(node.PersistentState.Log) {
+			nodeCommitIndex = len(node.PersistentState.Log)
+		}
+		node.VolatileState.CommitIndex = nodeCommitIndex
+		// apply the log command & update lastApplied
```

Comment:
Is this a TODO?

Reply:
Yeah, in future when we run our SQL commands on database then we update lastApplied value. This comment is to remind where to put the logic for that

Reply:
Sure. Then make sure theyre of the form `// TODO: `

Reply:
okay

Reply:
If you want this merged, create a ticket for this todo and mention it in the comment. Otherwise, TODOs must not be merged!

Reply:
For this Todo, we currently haven't created what we will do for the logs when they would be applied to the state machine. That's why I have created the comment as a placeholder. I will create a feature request for this **Todo** but when we start working on storage engine we will update this with relevant function.

Reply:
that API is complete. Call `server.onReplication`, at least I think it's called that. @SUMUKHA-PK ?

Reply:
Yes but lets wait on merging this PR. Till then @Abby3017 will be writing tests for this until I can figure some stuff out :D